lib/ansible/plugins/lookup/template.py

```diff
@@ -107,6 +107,9 @@ def run(self, terms, variables, **kwargs):
                 # do the templating
                 res = self._templar.template(template_data, preserve_trailing_newlines=True,
                                              convert_data=convert_data_p, escape_backslashes=False)
+                # restore old variables
+                self._templar.set_available_variables(variables)
+
```

Comment:
i would push this to the top of the loop

Reply:
How would that work? We are setting lookup file/template specific vars on line 105, before templating happens. So the restore needs to happen right after templating.

Reply:
Sorry, i was unclear, I didn't mean just this line, meant the vars copy/update block . 

Revisiting this line, I would actually put outside loop, after it ends or even change this to 'get_variables/set_variables' and avoid the ones you are passing in.

Also the existing `variables.copy()` will bypass your efforts as 'container types' will copy by ref, making modifications persist across not only invocations but in returned vars. see deepcopy or deepish_copy.

Reply:
> Sorry, i was unclear, I didn't mean just this line, meant the vars copy/update block .

I don't follow, what would be the gain if we moved the vars copy/update block to the top of the loop?

> Revisiting this line, I would actually put outside loop, after it ends.

> Also the existing variables.copy() will bypass your efforts as 'container types' will copy by ref, making modifications persist across not only invocations but in returned vars. see deepcopy or deepish_copy.

Good points, fixed both.

Reply:
the moving is more stylistic (prep on top), so you can ignore